design_documents/parametrizing_tests_with_fixed_values.md

```diff
@@ -0,0 +1,168 @@
+# Parametrizing Tests With Fixed Values
+
+## Context
+
+Various test runners for other programming languages provide ways of parametrizing test cases with fixed values.
+This way, multiple test cases can be executed for the same test source code.
+
+## Goal
+
+Propose a way of adding support for parametrized tests to Starknet Foundry's `snforge`.
+
+## Proposed Solution
+
+Introduce two new attributes: `parametrize` and `case`.
+
+- `parametrize` - defines a parametrized test.
+- `case` - defines a specific test case
+
+### Parametrization With Simple Arguments
+
+```cairo
+#[parametrize]
+#[case(1, 3)]
+#[case(3, 5)]
+#[test]
+fn my_test(a: felt252, b: u32) {
+    // case a = 1, b = 3
+    // case a = 3, b = 5
+    // ...
+}
+```
+
+For test cases like this, values could simply be injected into a runner for each `case`.
+
+### Parametrization With Fixtures
+
+```cairo
+struct MyStruct {
+    a: felt252, 
+    b: Array<felt252>
+}
+
+fn my_fixture() -> MyStruct {
+    MyStruct { a: 1, b: array![2, 3] }
+}
+
+#[parametrize]
+#[case(1, my_fixture())]
+#[case(3, my_fixture())]
+#[test]
+fn my_test(a: felt252, b: MyStruct) {
+    // case a = 1, b = MyStruct { a: 1, b: [2, 3] }
+    // case a = 3, b = MyStruct { a: 1, b: [2, 3] }
+    // ...
+}
+
+// This would require some code generation
+```
+
+To handle parametrization with function calls, we would have to generate a necessary code.
+An example generated code for the snippet above could look like this
+
+```cairo
+// ...
+#[test]
+fn my_test_generated(a: felt252) {
+    let b = my_fixture();
+    // ...
+}
+```
```

Comment:
So if I have this test
```
#[case(2, 1, my_fixture(), 7]
fn my_test(a: u8, b: u8, c: MyStruct, d: u8)
```

my generated code would look like
```
fn my_test_generated(a: u8, b: u8, d: u8) {
    let c = my_fixture();
```
right?

Reply:
Correct. There's some ambiguity in how "removing" arguments from function signature would be handled, so details of this may vary. But that was my idea.

Reply:
Yeah, that was my point. It seems complicated

Reply:
Some complication could maybe be removed if we decide to rely entirely on code generation: Each case would generate a new cairo code with all arguments inlined. 

It would cost us some extra code, but I'm assuming users would most often define only a couple of cases.

The problem with this approach would be ensuring some determinism in how these cases are executed: I'd like parametrized tests cases to run in order they were defined, so you always get deterministic failures from them.

Reply:
Updated the code so we only rely on code generation.